usr/lib/cca_stdll/cca_specific.c

```diff
@@ -2858,8 +2882,22 @@ static CK_RV ccatok_pkey_skey2pkey(STDLL_TokData_t *tokdata,
     pkey_wrap_handler_data.pkey_buflen_p = &pkey_buflen;
     pkey_wrap_handler_data.keytype = key_type;
 
-    ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
-                               &pkey_wrap_handler_data);
+    while (num_retries < 3600) {
+
+        ret = cca_iterate_adapters(tokdata, ccatok_pkey_sec2prot,
+                                   &pkey_wrap_handler_data);
+
+        if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+
+            if (ccatok_pkey_token_is_encr_with_new_mk(tokdata, key_type, mkvp)) {
```

Comment:
Please add the call to `check_expected_mkvp()` above right after the call to `analyse_cca_key_token()` with the following error handling: 

```
    if (check_expected_mkvp(tokdata, keytype, mkvp, &new_mk) != CKR_OK) {
        TRACE_ERROR("%s\n", ock_err(ERR_DEVICE_ERROR));
        rc = CKR_DEVICE_ERROR;
        goto done;
    }
```

We do want to detect if the blob is using the right MKVP (old or new) in any case. 
You get the `new_mk` flag that way right away. You can just check for `new_mk == TRUE` inside the loop instead of calling `ccatok_pkey_mk_change_active()`

Also, instead of calling `ccatok_pkey_mk_change_active()`, please just use `cca_mk_change_find_op_by_keytype(...) != NULL`.
No need for adding any new functions here.

Reply:
experimentally done, will push again ...